sdk/storage/azure-storage-file-share/src/test/java/com/azure/storage/file/share/BuilderHelperTests.java

```diff
@@ -0,0 +1,568 @@
+            .subscribe(response -> assertEquals(200, response.getStatusCode()));
+    }
+
+    /**
+     * Tests that a new date will be applied to every retry when using the share client builder's default pipeline.
+     */
+    @Test
+    public void shareClientFreshDateOnRetry() throws MalformedURLException {
+        ShareClient shareClient = new ShareClientBuilder()
+            .endpoint(ENDPOINT)
+            .shareName("share")
+            .credential(CREDENTIALS)
+            .httpClient(new FreshDateTestClient())
+            .retryOptions(REQUEST_RETRY_OPTIONS)
+            .buildClient();
+
+        shareClient.getHttpPipeline().send(request(shareClient.getShareUrl()))
+            .subscribe(response -> assertEquals(200, response.getStatusCode()));
+    }
+
+    /**
+     * Tests that a new date will be applied to every retry when using the file client builder's default pipeline.
+     */
+    @Test
+    void fileClientFreshDateOnRetry() throws MalformedURLException {
+        ShareFileClientBuilder fileClientBuilder = new ShareFileClientBuilder()
+            .endpoint(ENDPOINT)
+            .shareName("fileSystem")
+            .resourcePath("path")
+            .credential(CREDENTIALS)
+            .retryOptions(REQUEST_RETRY_OPTIONS)
+            .httpClient(new FreshDateTestClient());
+
+        ShareDirectoryClient directoryClient = fileClientBuilder.buildDirectoryClient();
+        directoryClient.getHttpPipeline().send(request(directoryClient.getDirectoryUrl()))
+            .subscribe(response -> assertEquals(200, response.getStatusCode()));
+
+        ShareFileClient fileClient = fileClientBuilder.buildFileClient();
+        fileClient.getHttpPipeline().send(request(fileClient.getFileUrl()))
+            .subscribe(response -> assertEquals(200, response.getStatusCode()));
+    }
+
+    /**
+     * Tests that a user application id will be honored in the UA string when using the default pipeline builder.
+     */
+    @ParameterizedTest
+    @MethodSource("customApplicationIdInUAStringSupplier")
+    public void customApplicationIdInUAString(List<String> applicationIds) throws MalformedURLException {
+        String logOptionsUA = applicationIds.get(0);
+        String clientOptionsUA = applicationIds.get(1);
+        String expectedUA = applicationIds.get(2);
+        HttpClient httpClient = new ApplicationIdUAStringTestClient(expectedUA);
+        HttpLogOptions httpLogOptions = new HttpLogOptions().setApplicationId(logOptionsUA);
+        ClientOptions clientOptions = new ClientOptions().setApplicationId(clientOptionsUA);
+
+        BuilderHelper.buildPipeline(CREDENTIALS, null, null, null, null, new RequestRetryOptions(), null,
+                httpLogOptions, clientOptions, httpClient, new ArrayList<>(), new ArrayList<>(),
+                Configuration.NONE, new ClientLogger("foo"))
+            .send(request(ENDPOINT))
+            .as(StepVerifier::create)
```

Comment:
Generally, we've used `StepVerifier.create(Mono/Flux)` instead of this `as` pattern. For consistency across the repo I'd recommend changing this, but this isn't critical.